juju/juju
# Feature remove space

### Checklist


----

## Description of change

- fix intermittent constraintvalidation test failure
- apiserver changes for the remove-space command
- new apiserver endpoint `remove-space` which takes a string to remove a space
  - before a space can be removed it will be checked on multiple places whether a remove is possible or not
  - do not remove on: 
      - existing bindings
      - if alpha space
      - existing model constraints and application constraints
- on deletion subnets will be moved back to alpha space
- cmd patch will follow to finish up this work

the further spec can be read here: https://docs.google.com/document/d/1tE5GMF9Uw8W7QQoybgA_vs_RqCHQdbKXmQT_Kaaj6Pk/edit#heading=h.leed10khee7o

## QA steps

- The cmd part is not implemented yet. Therefore a follow-up patch is needed.
- To be able to test this + cmd patch, one can check out the following example PR:
https://github.com/nammn/juju/pull/3
Especially apply this patch
https://github.com/nammn/juju/commit/644c2469a3ee2c578da451cd2bc6cb03b254251a.diff


Using the above PR the output should lead to the following.
```
❯ juju remove-space alpha 
ERROR cannot remove space "alpha": Could not remove space: "alpha"
Found the following constraints: [{"application-mysql"}]
Found the following bindings: [{"application-mysql"}]

~
❯ juju spaces            
Space  Name     Subnets       
3      aaabbbb                
0      alpha    172.31.1.0/28 
                172.31.16.0/20
                172.31.2.0/28 
                172.31.3.0/24 
                252.1.0.0/20  
                252.16.0.0/12 
                252.2.0.0/20  
                252.3.0.0/16  
1      bla      172.31.0.0/28 
                252.0.0.0/20  
                              
~
❯ juju get-model-constraints
spaces=bla

~
❯ juju remove-space bla     
ERROR cannot remove space "bla": Could not remove space: "bla"
Found the following constraints: [{"model-e"} {"application-mysql"}]


~
❯ juju show-application mysql              
mysql:
  charm: mysql
  series: xenial
  channel: stable
  constraints:
    spaces:
    - alpha
    - bla
  principal: true
  exposed: false
  remote: false
  endpoint-bindings:
    "": alpha
    ceph: alpha
...
~
❯ 

```
